### ISSUE TITLE
关于所提供预训练模型与test.py的模型尺寸不对应的问题

### ISSUE THREAD

**Issue by 1144181135** [kaiwang960112/Self-Cure-Network#56]
你好，我从https://drive.google.com/drive/folders/1dpDntw-2jLCNgjZ20W9aLg8a6DEEo7gX下载的预训练模型，可视化之后发现state_dict的key和value shape与原模型参数对应不上，如下所示。与class Res18Feature(nn.Module)对不上，导致无法加载，请问这是什么原因？
预训练模型key，value shape：
conv1.weight torch.Size([64, 3, 7, 7])
bn1.weight torch.Size([64])
bn1.bias torch.Size([64])
bn1.running_mean torch.Size([64])
bn1.running_var torch.Size([64])
layer1.0.conv1.weight torch.Size([64, 64, 3, 3])
layer1.0.bn1.weight torch.Size([64])
layer1.0.bn1.bias torch.Size([64])
layer1.0.bn1.running_mean torch.Size([64])
layer1.0.bn1.running_var torch.Size([64])
layer1.0.conv2.weight torch.Size([64, 64, 3, 3])
layer1.0.bn2.weight torch.Size([64])
layer1.0.bn2.bias torch.Size([64])
layer1.0.bn2.running_mean torch.Size([64])
layer1.0.bn2.running_var torch.Size([64])
layer1.1.conv1.weight torch.Size([64, 64, 3, 3])
layer1.1.bn1.weight torch.Size([64])
layer1.1.bn1.bias torch.Size([64])
layer1.1.bn1.running_mean torch.Size([64])
layer1.1.bn1.running_var torch.Size([64])
layer1.1.conv2.weight torch.Size([64, 64, 3, 3])
layer1.1.bn2.weight torch.Size([64])
layer1.1.bn2.bias torch.Size([64])
layer1.1.bn2.running_mean torch.Size([64])
layer1.1.bn2.running_var torch.Size([64])
layer2.0.conv1.weight torch.Size([128, 64, 3, 3])
layer2.0.bn1.weight torch.Size([128])
layer2.0.bn1.bias torch.Size([128])
layer2.0.bn1.running_mean torch.Size([128])
layer2.0.bn1.running_var torch.Size([128])
layer2.0.conv2.weight torch.Size([128, 128, 3, 3])
layer2.0.bn2.weight torch.Size([128])
layer2.0.bn2.bias torch.Size([128])
layer2.0.bn2.running_mean torch.Size([128])
layer2.0.bn2.running_var torch.Size([128])
layer2.0.downsample.0.weight torch.Size([128, 64, 1, 1])
layer2.0.downsample.1.weight torch.Size([128])
layer2.0.downsample.1.bias torch.Size([128])
layer2.0.downsample.1.running_mean torch.Size([128])
layer2.0.downsample.1.running_var torch.Size([128])
layer2.1.conv1.weight torch.Size([128, 128, 3, 3])
layer2.1.bn1.weight torch.Size([128])
layer2.1.bn1.bias torch.Size([128])
layer2.1.bn1.running_mean torch.Size([128])
layer2.1.bn1.running_var torch.Size([128])
layer2.1.conv2.weight torch.Size([128, 128, 3, 3])
layer2.1.bn2.weight torch.Size([128])
layer2.1.bn2.bias torch.Size([128])
layer2.1.bn2.running_mean torch.Size([128])
layer2.1.bn2.running_var torch.Size([128])
layer3.0.conv1.weight torch.Size([256, 128, 3, 3])
layer3.0.bn1.weight torch.Size([256])
layer3.0.bn1.bias torch.Size([256])
layer3.0.bn1.running_mean torch.Size([256])
layer3.0.bn1.running_var torch.Size([256])
layer3.0.conv2.weight torch.Size([256, 256, 3, 3])
layer3.0.bn2.weight torch.Size([256])
layer3.0.bn2.bias torch.Size([256])
layer3.0.bn2.running_mean torch.Size([256])
layer3.0.bn2.running_var torch.Size([256])
layer3.0.downsample.0.weight torch.Size([256, 128, 1, 1])
layer3.0.downsample.1.weight torch.Size([256])
layer3.0.downsample.1.bias torch.Size([256])
layer3.0.downsample.1.running_mean torch.Size([256])
layer3.0.downsample.1.running_var torch.Size([256])
layer3.1.conv1.weight torch.Size([256, 256, 3, 3])
layer3.1.bn1.weight torch.Size([256])
layer3.1.bn1.bias torch.Size([256])
layer3.1.bn1.running_mean torch.Size([256])
layer3.1.bn1.running_var torch.Size([256])
layer3.1.conv2.weight torch.Size([256, 256, 3, 3])
layer3.1.bn2.weight torch.Size([256])
layer3.1.bn2.bias torch.Size([256])
layer3.1.bn2.running_mean torch.Size([256])
layer3.1.bn2.running_var torch.Size([256])
layer4.0.conv1.weight torch.Size([512, 256, 3, 3])
layer4.0.bn1.weight torch.Size([512])
layer4.0.bn1.bias torch.Size([512])
layer4.0.bn1.running_mean torch.Size([512])
layer4.0.bn1.running_var torch.Size([512])
layer4.0.conv2.weight torch.Size([512, 512, 3, 3])
layer4.0.bn2.weight torch.Size([512])
layer4.0.bn2.bias torch.Size([512])
layer4.0.bn2.running_mean torch.Size([512])
layer4.0.bn2.running_var torch.Size([512])
layer4.0.downsample.0.weight torch.Size([512, 256, 1, 1])
layer4.0.downsample.1.weight torch.Size([512])
layer4.0.downsample.1.bias torch.Size([512])
layer4.0.downsample.1.running_mean torch.Size([512])
layer4.0.downsample.1.running_var torch.Size([512])
layer4.1.conv1.weight torch.Size([512, 512, 3, 3])
layer4.1.bn1.weight torch.Size([512])
layer4.1.bn1.bias torch.Size([512])
layer4.1.bn1.running_mean torch.Size([512])
layer4.1.bn1.running_var torch.Size([512])
layer4.1.conv2.weight torch.Size([512, 512, 3, 3])
layer4.1.bn2.weight torch.Size([512])
layer4.1.bn2.bias torch.Size([512])
layer4.1.bn2.running_mean torch.Size([512])
layer4.1.bn2.running_var torch.Size([512])
feature.weight torch.Size([256, 512])
feature.bias torch.Size([256])
fc.weight torch.Size([87020, 256])
fc.bias torch.Size([87020])

**Comment by kaiwang960112** [comment]
这个回答过了，是key不一致的问题，相差了个module.

1144181135 ***@***.***>于2021年7月21日 周三上午11:59写道：

**Comment by 1144181135** [comment]
你好，我已经把预训练模型的module.去掉了，上面已经列出来了我去掉module.后的key和value shape。这个issue中列的是预训练模型的key和value shape。首先，class Res18Feature(nn.Module) 中没有上述key。第二，self.features应该是一个backbone，但在预训练模型的参数中feature对应的是一个Linear参数。第三，预训练模型中出现了{fc.weight torch.Size([87020, 256])，fc.bias torch.Size([87020])}，这个参数尺寸在class Res18Feature(nn.Module):中并没有找到。

class Res18Feature(nn.Module):
    def __init__(self, pretrained, num_classes = 7):
        super(Res18Feature, self).__init__()
        resnet  = models.resnet18(pretrained)
        # self.feature = nn.Sequential(*list(resnet.children())[:-2]) # before avgpool
        self.features = nn.Sequential(*list(resnet.children())[:-1]) # after avgpool 512x1

        fc_in_dim = list(resnet.children())[-1].in_features # original fc layer's in dimention 512

        self.fc = nn.Linear(fc_in_dim, num_classes) # new fc layer 512x7
        self.alpha = nn.Sequential(nn.Linear(fc_in_dim, 1),nn.Sigmoid())

**Comment by dzz416** [comment]
求大佬说一下解决了没，怎么解决的，我是小白，想跑一下测试都跑不通

**Comment by dreamlychina** [comment]
> 

人家问的是fc层的参数和你给的对应不上的问题，这代码真的太垃圾了

**Comment by kaiwang960112** [comment]
我不太理解你为啥finetune的时候还要人脸预训练模型的分类器。。。。。。。，人脸预训练模型的分类器的类别数和表情分类类别数都不一样。你load干啥？

**Comment by kaiwang960112** [comment]
> 你好，我已经把预训练模型的module.去掉了，上面已经列出来了我去掉module.后的key和value shape。这个issue中列的是预训练模型的key和value shape。首先，class Res18Feature(nn.Module) 中没有上述key。第二，self.features应该是一个backbone，但在预训练模型的参数中feature对应的是一个Linear参数。第三，预训练模型中出现了{fc.weight torch.Size([87020, 256])，fc.bias torch.Size([87020])}，这个参数尺寸在class Res18Feature(nn.Module):中并没有找到。
> 
> class Res18Feature(nn.Module): def **init**(self, pretrained, num_classes = 7): super(Res18Feature, self).**init**() resnet = models.resnet18(pretrained) # self.feature = nn.Sequential(*list(resnet.children())[:-2]) # before avgpool self.features = nn.Sequential(*list(resnet.children())[:-1]) # after avgpool 512x1
> 
> ```
>     fc_in_dim = list(resnet.children())[-1].in_features # original fc layer's in dimention 512
> 
>     self.fc = nn.Linear(fc_in_dim, num_classes) # new fc layer 512x7
>     self.alpha = nn.Sequential(nn.Linear(fc_in_dim, 1),nn.Sigmoid())
> ```

**Comment by dreamlychina** [comment]
> 

你放个能跑的模型会死啊，代码写的烂还不让人说。那类别数不一样，前面的名字都不一样，去掉module也不能加载

**Comment by kaiwang960112** [comment]
好的，谢谢建议


dreamlychina ***@***.***>于2022年4月19日 周二下午4:53写道：
